packages/bson-transpilers/codegeneration/CodeGenerationVisitor.js

```diff
@@ -53,15 +55,32 @@ module.exports = (ANTLRVisitor) => class CodeGenerationVisitor extends ANTLRVisi
     return this[rule](ctx);
   }
 
+  returnResultWithDeclarations(ctx) {
+    let result = this.returnResult(ctx);
+    if (this.getState().declarations.length() > 0) {
+      result = `${this.getState().declarations.toString() + '\n\n'}${result}`;
+    }
+    return result;
+  }
+
   /**
    * PUBLIC: This is the entry point for the compiler. Each visitor must define
    * an attribute called "startNode".
    *
    * @param {ParserRuleContext} ctx
+   * @param {Boolean} useDeclarations - prepend the result string with declarations
    * @return {String}
    */
-  start(ctx) {
-    return this.returnResult(ctx).trim();
+  start(ctx, useDeclarations = false) {
+    return (useDeclarations ? this.returnResultWithDeclarations(ctx) : this.returnResult(ctx)).trim();
```

Comment:
This change is definitely of note, it's what allows us to prepend the raw document output with functions/variables:

<img width="894" alt="Screen Shot 2022-04-13 at 11 41 22 AM" src="https://user-images.githubusercontent.com/24281431/163276391-4fc08d9f-8a63-4b39-b611-54cf217cd8db.png">

There might be a much better way of going about this, though.